### ISSUE TITLE
Future request, diff on save / diff on upload. 

### ISSUE THREAD

**Issue by AdamJonsson** [liximomo/vscode-sftp#215]
A new future what many would appreciate is the option to diff on upload. A future that more or less checks if the file on the server have changes after the project was dowloaded. 

Maybe it could be done by, when saving, downloading the file that is going to be saved from the server. Compare it with the local one (Before saving the file). If there is no difference between the files, save the file and upload it. If there is a difference, cancel the save and show a warning and the diff panel between the local and the remote file.

**Comment by kaidoj** [comment]
Went back to phpstorm just because of this. Im waiting this to be added and im back on vscode. Please add! Thank you.

**Comment by liximomo** [comment]
@AdamJonsson Does phpstorm support this?

**Comment by AdamJonsson** [comment]
@liximomo Yes they do. It is located under: Default Preferences -> Build, Execution, Deployment -> Deployment -> Options. There they have a setting that is called: "Warn when uploading over newer file" and also "Notify of remote changes". 

![image](https://user-images.githubusercontent.com/13873828/42983924-d49322b6-8be9-11e8-9f3b-1df55dccdc56.png)

However, I have no idea how they are doing it. An easy fix may just to compare the timestamp or size.

**Comment by Jeffbabb5** [comment]
I would love this feature as well.

Thanks!

**Comment by dadavec** [comment]
Any update about this feature? I believe you already have all neded parts ready, so we would just need an option like "diffOnUpload". 

It would run the diff command you already have before the upload. And then in case files will be same upload will be done, otherwise it will fail/stop with warning.

**Comment by AndryWJ** [comment]
in ftp-kr it is. Only broke after updating vscode. Very useful thing!

**Comment by dadavec** [comment]
I was using it before, but since it broke up I was looking for some alternative. This one is very close and this feature is the only one I am missing.

**Comment by ildarkhasanshin** [comment]
> in ftp-kr it is. Only broke after updating vscode. Very useful thing!

+
only with option "autoDownload": true

**Comment by Lospepes** [comment]
Detecting changes automatically between local and remote is missing to this extension.  Is it possible one day ? 

And a configuration to automatically download from remote an open file from the local explorer would be a real advantage.

Every SFTP extension is missing something major like this.

If someone guide me, maybe I could try to do it ?

**Comment by cihankaba** [comment]
> Detecting changes automatically between local and remote is missing to this extension. Is it possible one day ?
> 
> And a configuration to automatically download from remote an open file from the local explorer would be a real advantage.
> 
> Every SFTP extension is missing something major like this.
> 
> If someone guide me, maybe I could try to do it ?

Absolutely I agree. I use SASS extension and when save style.scss file, i have to save style.css file too.

**Comment by Pudelwerfer** [comment]
+1, diff before upload on save before overwrite newer files 

Even Dreamweaver has it built in. (In combination with an external tool like WinMerge)

I come from HomeSite and have been using DreamWeaver for pure coding (yes;) ever since and I would very much like to switch to VSC. This is one of two features that I sorely miss compared to the DW. (Or I haven't found it yet.)

Here newer files are completely overwritten without warning. **This is dangerous.** Why compare manually every time before the upload when you can automate it?

If the remote file is newer, then at least ask for "diff or overwrite"

MJ

**Comment by gimjudge** [comment]
+1 I'm simultaneously using using your stfp extension and ftp-kr. Close to disabling this extension. ftp-kr throws the diff on the screen and gives the option "upload" or "download" when there is a difference when using autoUpload

**Comment by SomeDevWeb** [comment]
Hi guys. Any news about this?
I'm also very keen on having it. I recently had some issues where I destroyed a few important code updates I did on another laptop and I forgot about them, so it's quite vital in some cases.
